File: src/lib/reserved.ts
/**
 * Reserved-word list — single source for both Next route definitions and alias
 * validation (ARCHITECTURE.md §3.1, FR-3, AC-5). Kept in code so routing and
 * validation can never drift. Comparison is case-insensitive.
 */
export const RESERVED_WORDS: readonly string[] = [
  'api',
  'login',
  'signin',
  'signup',
  'logout',
  'app',
  'dashboard',
  'admin',
  'settings',
  'account',
  'analytics',
  'links',
  'bulk',
  'qr',
  'auth',
  'healthz',
  'health',
  'dead-link',
  'gate',
  '_next',
  'static',
  'assets',
  'favicon.ico',
  'robots.txt',
  'sitemap.xml',
] as const

const RESERVED_SET = new Set(RESERVED_WORDS.map((w) => w.toLowerCase()))

/** True if `value` collides with a reserved app route (case-insensitive). */
export function isReserved(value: string): boolean {
  return RESERVED_SET.has(value.trim().toLowerCase())
}

File: backend/src/services/reserved.rs
//! Reserved-word list — ported from `src/lib/reserved.ts`.
//! Single source for routing and alias validation. Comparison is
//! case-insensitive.

/// Reserved words that collide with app routes and cannot be used as aliases.
pub const RESERVED_WORDS: &[&str] = &[
    "api",
    "login",
    "signin",
    "signup",
    "logout",
    "app",
    "dashboard",
    "admin",
    "settings",
    "account",
    "analytics",
    "links",
    "bulk",
    "qr",
    "auth",
    "healthz",
    "health",
    "dead-link",
    "gate",
    "_next",
    "static",
    "assets",
    "favicon.ico",
    "robots.txt",
    "sitemap.xml",
];

/// True if `word` collides with a reserved app route (case-insensitive).
pub fn is_reserved(word: &str) -> bool {
    let normalized = word.trim().to_lowercase();
    RESERVED_WORDS.iter().any(|w| *w == normalized)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn matches_case_insensitively() {
        assert!(is_reserved("api"));
        assert!(is_reserved("API"));
        assert!(is_reserved("  Admin  "));
        assert!(is_reserved("favicon.ico"));
        assert!(is_reserved("dead-link"));
    }

    #[test]
    fn allows_non_reserved() {
        assert!(!is_reserved("my-link"));
        assert!(!is_reserved("apix"));
        assert!(!is_reserved("hello"));
    }
}
